src/SampleLibraryUI/Examples/ButtonCustomNodeModel.cs

```diff
@@ -0,0 +1,336 @@
+        /// </summary>
+        [JsonIgnore]
+        [IsVisibleInDynamoLibrary(false)]
+        public DelegateCommand ButtonCommand { get; set; }
+
+        #endregion
+
+        #region constructor
+
+        /// <summary>
+        /// The constructor for a NodeModel is used to create
+        /// the input and output ports and specify the argument
+        /// lacing.
+        /// </summary>
+        public ButtonCustomNodeModel()
+        {
+            // When you create a UI node, you need to do the
+            // work of setting up the ports yourself. To do this,
+            // you can populate the InPorts and the OutPorts
+            // collections with PortData objects describing your ports.
+            InPorts.Add(new PortModel(PortType.Input, this, new PortData("inputString", "a string value displayed on our button")));
+
+            // Nodes can have an arbitrary number of inputs and outputs.
+            // If you want more ports, just create more PortData objects.
+            OutPorts.Add(new PortModel(PortType.Output, this, new PortData("button value", "returns the string value displayed on our button")));
+            OutPorts.Add(new PortModel(PortType.Output, this, new PortData("window value", "returns the string value displayed in our window when button is pressed")));
+
+            // This call is required to ensure that your ports are
+            // properly created.
+            RegisterAllPorts();
+
+            // Listen for input port disconnection to trigger button UI update
+            this.PortDisconnected += ButtonCustomNodeModel_PortDisconnected;
+
+            // The arugment lacing is the way in which Dynamo handles
+            // inputs of lists. If you don't want your node to
+            // support argument lacing, you can set this to LacingStrategy.Disabled.
+            ArgumentLacing = LacingStrategy.Disabled;
+
+            // We create a DelegateCommand object which will be 
+            // bound to our button in our custom UI. Clicking the button 
+            // will call the ShowMessage method.
+            ButtonCommand = new DelegateCommand(ShowMessage, CanShowMessage);
+
+            // Setting our property here will trigger a 
+            // property change notification and the UI 
+            // will be updated to reflect the new value.
+            ButtonText = defaultButtonText;
+            WindowText = defaultWindowText;
+        }
+
+        // Starting with Dynamo v2.0 you must add Json constructors for all nodeModel
+        // dervived nodes to support the move from an Xml to Json file format.  Failing to
+        // do so will result in incorrect ports being generated upon serialization/deserialization.
+        [JsonConstructor]
+        ButtonCustomNodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts)
+        {
+            this.PortDisconnected += ButtonCustomNodeModel_PortDisconnected;
```

Comment:
@mjkkirschner I should be unsubscribing from this in my dispose method?

Reply:
this actually should not be required, because the method that is subscribed is actually on the same object that is raising the event so theres no reference to the object except from itself.

Reply:
no harm in doing it just in case though.